tests/core/pyspec/eth2spec/test/bellatrix/sync/test_optimistic.py

```diff
@@ -112,3 +114,209 @@ def test_from_syncing_to_invalid(spec, state):
+    yield from add_optimistic_block(spec, mega_store, signed_block, test_steps, status=PayloadStatusV1Status.VALID)
+    assert spec.get_head(mega_store.fc_store) == mega_store.opt_store.head_block_root
+
+    # Create SYNC chains
+    state_0 = state.copy()
+    aggregation_bit_lists = [[1, 0, 0 ,0], [0, 1, 0 ,0], [0, 0, 1, 0]]
+    for j, level in enumerate(["a", "b", "c"]):
+        state = state_0.copy()
+        for i in range(3):
+            print(f"{level}_{i}")
+            block = build_empty_block_for_next_slot(spec, state)
+            block.body.execution_payload.parent_hash = (
+                block_hashes[f'chain_{level}_{i - 1}'] if i != 0 else block_hashes['block_0']
+            )
+            block.body.execution_payload.extra_data = spec.hash(bytes(f'chain_{level}_{i}', 'UTF-8'))
+            block.body.execution_payload.block_hash = compute_el_block_hash(spec, block.body.execution_payload)
+            block_hashes[f'chain_{level}_{i}'] = block.body.execution_payload.block_hash
+
+            max_committee_index = spec.get_committee_count_per_slot(state, current_slot)
+            committee_index = min(j, max_committee_index-1)
+            signed_block = add_attestation_and_sign_block_with_aggregation_bit_list(
+                spec, state, fc_store, block, committee_index, aggregation_bit_lists[j]
+            )
+            signed_blocks[f'chain_{level}_{i}'] = signed_block.copy()
+
+            yield from add_optimistic_block(spec, mega_store, signed_block, test_steps,
+                                            status=PayloadStatusV1Status.SYNCING)
+            assert spec.get_head(mega_store.fc_store) == mega_store.opt_store.head_block_root
+        state_store[level] = state.copy()
+
+    # Last added branch is considered as optimistic head
+    assert mega_store.opt_store.head_block_root == spec.Root(signed_blocks[f'chain_c_2'].message.hash_tree_root())
+
+    latest_valid_hash = block_0.body.execution_payload.block_hash
+
+    # Add an invalid block to chain C
+    block = build_empty_block_for_next_slot(spec, state_store["c"])
+    block.body.execution_payload.parent_hash = signed_blocks[f'chain_c_2'].message.body.execution_payload.block_hash
+    block.body.execution_payload.extra_data = spec.hash(bytes(f'chain_c_3', 'UTF-8'))
+    block.body.execution_payload.block_hash = compute_el_block_hash(spec, block.body.execution_payload)
+    signed_block = state_transition_and_sign_block(spec, state_store["c"], block)
+    payload_status = PayloadStatusV1(
+        status=PayloadStatusV1Status.INVALID,
+        latest_valid_hash=latest_valid_hash,
+        validation_error="invalid",
+    )
+    yield from add_optimistic_block(spec, mega_store, signed_block, test_steps,
+                                    payload_status=payload_status)
+
+    assert mega_store.opt_store.head_block_root == spec.Root(signed_blocks[f'chain_b_2'].message.hash_tree_root())
```

Comment:
In this case ties should be broken lexicographically like specified in the [`get_head`](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/fork-choice.md#get_head), right? So for this assertion to pass `chain_b_2` block's root should be greater than `chain_a_2`'s one. Just confirming.

Reply:
That's what I think too.

If I save each signed block root (I store this in a dict `spec.Root(signed_blocks[f'chain_{level}_{i}'].message.hash_tree_root())`), I have the below:

```
{"chain_a_0": 0x0e2562911c730a717495bddfa3e88dd820280b8a75c03f833d3e72fe623cc746,
 "chain_a_1": 0x325fc42cdc5af6fd011bcf83aca258d64b397b4477e99b40451ff55824a61068,
 "chain_a_2": 0x0aaa43953f2c9e61b28f8dda56c4171ba334af19ff036f706195c4baa8463240,
 "chain_b_0": 0x44bc074aeff8705acc3dada7173c893efa2136f143779ee7ed9cded60a44e256,
 "chain_b_1": 0x0500a1c275c3d44141851c630009c6a4723b22735f4e88ac32d019504c5b032a,
 "chain_b_2": 0xdbc5f3f1076df87eed0b4111094b24351ec19bd858edc014869429d82e733bbf,
 "chain_c_0": 0xb9dc73fc6ca54026dfe47f05871bcc0960788d7bbc5a2846e574cdd6797a9000,
 "chain_c_1": 0xacbf6b3462d83504a0e6751d20a2ac55284b1f5dbe8630c70b69c46faf76aeee,
 "chain_c_2": 0x5e616b8a71287fc957e1e6ab3e14c3613bccf7a3bf4fc8c1ca81f1e47aaf415c}
```

What define what is the optimistic head is not `get_head` as you said but a very similar method [`get_opt_head_block_root`](https://github.com/ethereum/consensus-specs/blob/e3a939e439d6c05356c9c29c5cd347384180bc01/tests/core/pyspec/eth2spec/test/helpers/optimistic_sync.py#L158). When we invalid the branch C, in the method `get_opt_head_block_root` we have this iteration pairs (root/weight):

1. 0xe5177eaeb2ab02baed986a20ed5decc9e0d60e7317cbf20e8503cd43d6beed88: 288000000000
2.
    0x0e2562911c730a717495bddfa3e88dd820280b8a75c03f833d3e72fe623cc746: 64000000000
    0x44bc074aeff8705acc3dada7173c893efa2136f143779ee7ed9cded60a44e256: 64000000000
3. 0x0500a1c275c3d44141851c630009c6a4723b22735f4e88ac32d019504c5b032a: 32000000000
4. 0xdbc5f3f1076df87eed0b4111094b24351ec19bd858edc014869429d82e733bbf: 0

As you can see, iteration 2 is the selection between branch A (0x0e2562911c730a717495bddfa3e88dd820280b8a75c03f833d3e72fe623cc746) and branch B (0x44bc074aeff8705acc3dada7173c893efa2136f143779ee7ed9cded60a44e256) => it seems we selected the branch B since the iteration 3 covered the branch B "chain_b_1" root.
That's because  0x44bc074aeff8705acc3dada7173c893efa2136f143779ee7ed9cded60a44e256 > 0x0e2562911c730a717495bddfa3e88dd820280b8a75c03f833d3e72fe623cc746.

To be sure I did test the below:
```python
def test_order(element):
    if element in ["a", "b"]:
        return 5
    return 1
```
>>> max(["x", "b"], key=lambda element: test_order(element))

Reply:
Exhaustive arguments! Thank you!